File: android/app/src/main/kotlin/com/goalmaker/app/domain/planning/PromptLibrary.kt
package com.goalmaker.app.domain.planning

import java.io.InputStream
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * The review prompts the app ships (contracts/content/prompts.json), in the file's order: the
 * rotation walks the [categories] and takes from [prompts] (docs/reviews.md).
 */
class PromptLibrary(val version: Int, val categories: List<String>, val prompts: List<ReviewPrompt>) {
    private val byId = prompts.associateBy(ReviewPrompt::id)

    operator fun get(id: String): ReviewPrompt? = byId[id]

    companion object {
        fun parse(text: String): PromptLibrary {
            val document = Json.parseToJsonElement(text).jsonObject
            return PromptLibrary(
                version = document.getValue("version").jsonPrimitive.content.toInt(),
                categories = document.getValue("categories").jsonArray.map { it.jsonObject.getValue("id").jsonPrimitive.content },
                prompts = document.getValue("prompts").jsonArray.map { element ->
                    val prompt = element.jsonObject
                    ReviewPrompt(
                        id = prompt.getValue("id").jsonPrimitive.content,
                        category = prompt.getValue("category").jsonPrimitive.content,
                        reviews = prompt.getValue("reviews").jsonArray.map { it.jsonPrimitive.content }.toSet(),
                        text = prompt.getValue("text").jsonPrimitive.content,
                        trigger = prompt.text("trigger"),
                    )
                },
            )
        }

        fun load(stream: InputStream): PromptLibrary = stream.use { parse(it.readBytes().toString(Charsets.UTF_8)) }

        private fun JsonObject.text(name: String): String? =
            this[name]?.jsonPrimitive?.takeIf { it.booleanOrNull == null && it.content != "null" }?.content
    }
}

File: windows/src/GoalMaker.Core/Planning/PromptLibrary.cs
using System.Text.Json;

namespace GoalMaker.Core.Planning;

/// <summary>
/// The review prompts the app ships (contracts/content/prompts.json), in the file's order: the rotation
/// walks the categories and takes from the prompts (docs/reviews.md).
/// </summary>
public sealed class PromptLibrary
{
    private readonly Dictionary<string, ReviewPrompt> byId;

    public PromptLibrary(int version, IReadOnlyList<string> categories, IReadOnlyList<ReviewPrompt> prompts)
    {
        Version = version;
        Categories = categories;
        Prompts = prompts;
        byId = prompts.ToDictionary(prompt => prompt.Id, StringComparer.Ordinal);
    }

    public int Version { get; }

    public IReadOnlyList<string> Categories { get; }

    public IReadOnlyList<ReviewPrompt> Prompts { get; }

    public ReviewPrompt? this[string id] => byId.TryGetValue(id, out var prompt) ? prompt : null;

    public static PromptLibrary Parse(string text)
    {
        using var document = JsonDocument.Parse(text);
        return Read(document.RootElement);
    }

    public static PromptLibrary Load(Stream stream)
    {
        using var document = JsonDocument.Parse(stream);
        return Read(document.RootElement);
    }

    private static PromptLibrary Read(JsonElement root) => new(
        root.GetProperty("version").GetInt32(),
        [.. root.GetProperty("categories").EnumerateArray().Select(category => category.GetProperty("id").GetString()!)],
        [.. root.GetProperty("prompts").EnumerateArray().Select(prompt => new ReviewPrompt(
            prompt.GetProperty("id").GetString()!,
            prompt.GetProperty("category").GetString()!,
            prompt.GetProperty("reviews").EnumerateArray().Select(kind => kind.GetString()!).ToHashSet(StringComparer.Ordinal),
            prompt.GetProperty("text").GetString()!,
            prompt.TryGetProperty("trigger", out var trigger) ? trigger.GetString() : null))]);
}
